### PR TITLE
MRG, ENH: Add volumetric atlas support

### LINKED ISSUE #6155: ENH: Volume labeling
At some point we should have a way of "label-izing" volumes like we do for surfaces. Freesurfer probably has utilities/files for this but I've never looked or worked with them (probably aseg.mgz), and in general we should have a way of making volume labels. Basically we need (API proposal):

1. `read_labels_from_aseg` (or whatever the correct Freesurfer file is), which returns a list of...
2. `VolumeLabel` or some equivalent (needs suitabl API and I/O functions)
3. `extract_label_time_course` extended to work for `VolSourecEstimate` + `VolLabel`

Eventually we should make everything work for mixed source spaces, too, but there are probably other things to sort out before then (e.g., mixed orientation priors).

### PR DESCRIPTION
**Pull request by larsoner**
Adds volume label support to `extract_label_time_course`. Requires #7597 (and follow-up PRs).

- [x] Add `mne.read_freesurfer_lut` as suggested by @vferat [here](https://github.com/mne-tools/mne-python/issues/6155#issuecomment-612420561) to make `setup_volume_source_space` more general and pave the way for more generic atlas support
- [x] Speed up `setup_volume_source_space` with many labels by refactoring `_make_volume_source_space`
- [x] Add test that `setup_volume_source_space` with all volume labels of `aseg.mgz` creates a complete source space
- [x] Add `extract_label_time_course` VolSourceEstimate support via atlas
- [x] Add `vol_stc.in_label` volume atlas support

Next PRs:

- Ensure all source estimate types work with SourceMorph (#7564)
- Update docs about `src` in SourceMorph (#7007)

Closes #6155.

### CODE REVIEW

**Comment by larsoner** [comment]
Okay @agramfort this is ready for review/merge as well. Might be good for you to take a first pass, then once we're happy with the API ask others who would be interested to look

**Comment by larsoner** [comment]
The new tutorial [image](https://19612-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/source-modeling/plot_visualize_stc.html#sphx-glr-auto-tutorials-source-modeling-plot-visualize-stc-py) showing `extract_label_time_course` on a volume STC:

File: mne/source_estimate.py [hunk 1]
```diff
@@ -1791,6 +1808,89 @@ def plot(self, src, subject=None, subjects_dir=None, mode='stat_map',
...
+        --------
+        extract_label_time_course : Extract time courses for multiple STCs.
+
+        Notes
+        -----
+        %(eltc_mode_notes)s
+        """
+        return extract_label_time_course(
+            self, labels, src, mode=mode, return_generator=False,
+            allow_empty=allow_empty, verbose=verbose)
```
Comment (agramfort) [inline mne/source_estimate.py:1845]: you don't pass `trans` or `mri_resolution` parameters here. I am confused

Comment (larsoner) [reply mne/source_estimate.py:1845]: Just a bug, need to write tests to use this method in addition to the function !

Comment (larsoner) [reply mne/source_estimate.py]: Fixed

**Comment by poojaprabhu9** [comment]
I was trying to read the volume stc which were computed previously using mne.read_source_estimate in mne.v0.21.dev0.
I met with an error
TypeError: vertices must be an instance of list, got <class 'numpy.ndarray'> instead.
without reading stcs i cannot check the extract_label_time_course ,correct?

**Comment by larsoner** [comment]
@poojaprabhu9 I pushed a fix that should fix that bug hopefully

**Comment by poojaprabhu9** [comment]
I tried the same by installing mne.v0.21.dev0. 
I error still exist.

**Comment by agramfort** [comment]
you need to use this branch. Try:

pip install -U https://github.com/larsoner/mne-python/archive/atlas.zip

**Comment by poojaprabhu9** [comment]
that helped, thanks.
I had computed stc after morphing.
Thus, now when i am extracting the time course, i have to setup the volume source space for  fsaverage  to get ''src" (same src was used while doing morphing)
```
src_fs = mne.setup_volume_source_space( 'fsaverage', pos=5.,mri=aseg_fname, 
             subjects_dir=subjects_dir,verbose=True)
```
then i tried to extract the time course from the stc

`tc=stc_morph.extract_label_time_course(labels='/path/fsaverage/mri/aparc+aseg.mgz',src=src_fs,mode='mean',allow_empty=True)`

i am encountering the unexpected error:
ValueError: 9900/20377 left hemisphere stc vertices missing from the source space, likely mismatch.

How is that possible when i am using the same src to extract the time course as well as to compute the morphed stc?

**Comment by agramfort** [comment]
can you share a code snippet to replicate ?

>

**Comment by poojaprabhu9** [comment]
I tried to replicate the code  [here](https://gist.github.com/poojaprabhu9/22e36e4d44d1d53da83fbec0c54ddd07)
two observations are done,
1. when i execute the code in mne.v.0.20,  the stc obtained after morphing has 20377 vertices whereas when executed in development version(pip install -U https://github.com/larsoner/mne-python/archive/atlas.zip ) it has 24303 vertices. I can extract the time course from the morphed stc computed from development version. In my case i have the morphed stc computed using mne.v.0.20, when i read this in development version and try to extract the time course i get the ValueError: 9900/20377 left hemisphere stc vertices missing from the source space, likely mismatch.
2.  when i try to execute (line 49 of gist)
`src = mne.setup_volume_source_space('sample', mri=aseg_fname,pos=5., sphere=sphere, mindist=3.,subjects_dir=subjects_dir)`
 in development version i get 
MemoryError: Unable to allocate 402. TiB for an array with shape (48003, 48003, 48003) and data type int32
whereas it works fine in mne.v.0.40.
so you can't run the whole code in development version. So i computed the forward solution in mne.v.0.20 and then i read the forward solution in development version.

**Comment by larsoner** [comment]
You have to be careful about sphere_units, they changed between versions. You probably need to set this to mm, the default is now m.

In general if you stick to using dev versions you need to follow changes or read the release notes very closely because deprecation cycles can be effectively days or weeks in duration instead of 6 months.

**Comment by poojaprabhu9** [comment]
> You have to be careful about sphere_units, they changed between versions. You probably need to set this to mm, the default is now m.
> 
> In general if you stick to using dev versions you need to follow changes or read the release notes very closely because deprecation cycles can be effectively days or weeks in duration instead of 6 months.

that solved the second point.
first problem is solved by adding sphere=(0,0,0,90) to 
```
src_fs = mne.setup_volume_source_space( 'fsaverage', pos=5.,mri=aseg_fname, 
             subjects_dir=subjects_dir,verbose=True)
```
Then compute the extract_time_course. 
I could extract the time course. Thanks!

**Comment by larsoner** [comment]
Okay for you @vferat ?

File: mne/source_space.py [hunk 2]
```diff
@@ -2535,36 +2615,37 @@ def get_volume_labels_from_aseg(mgz_fname, return_colors=False):
...
-                     lut[lut['id'] == ii]['A'][0]]
-                    for ii in np.unique(mgz_data)]
-
-    order = np.argsort(label_names)
-    label_names = [label_names[k] for k in order]
-    label_colors = [label_colors[k] for k in order]
-
+    atlas = nib.load(mgz_fname)
+    want = np.unique(_get_img_fdata(atlas))
+    ids, colors = read_freesurfer_lut()
```
Comment (vferat) [inline mne/source_space.py:2632]: Would it be possible to add a way to use specific freesurfer LUT here ? Adding an input either an object returned by or a file path to be read by `read_freesurfer_lut` :

`def get_volume_labels_from_aseg(mgz_fname, lut=None, return_colors=False):`

Comment (larsoner) [reply mne/source_space.py:2632]: So you want both `read_freesurfer_lut(fname=None)`, and `get_volume_labels_from_aseg(mgz_fname, return_colors=False, lut=None)` where non-None inputs mean "don't use the builtin one"? Seems doable

Comment (larsoner) [reply mne/source_space.py:2632]: Actually `read_freesurfer_lut` already takes `fname`, so we just need `lut` support for `get_volume_labels_from_aseg`

Comment (larsoner) [reply mne/source_space.py]: added `atlas_ids` argument in latest commit

**Comment by larsoner** [comment]
@vferat should be fixed now

**Comment by larsoner** [comment]
Oh wait I just fixed the error you were getting (pre-edit). Let me look...

**Comment by larsoner** [comment]
@vferat I get:
```
>>> len(mne.get_volume_labels_from_aseg('sample_Schaefer2018_100Parcels_7Networks.mgz'))
114
```
Maybe some bug with your code?

**Comment by larsoner** [comment]
Okay I see the problem. When you do only:
```
label_tc = stc.extract_label_time_course(fname_aseg, src=src, trans=inv['mri_head_t'])
```
It will only use the builtin FreeSurfer LUT with the `aseg` you have provided. You need to pass the `atlas_ids` you want to use with this non-standard MRI, e.g.:
```
atlas_ids, lut_colors = mne.read_freesurfer_lut('Schaefer2018_100Parcels_7Networks_LUT.txt')
label_names = mne.get_volume_labels_from_aseg(fname_aseg, atlas_ids=atlas_ids)
# not sure if you need this, but just to be safe:
atlas_ids = {k: atlas_ids[k] for k in label_names}
label_tc = stc.extract_label_time_course((fname_aseg, atlas_ids), src=src, trans=inv['mri_head_t'])
```

**Comment by vferat** [comment]
> 
> 
> Okay I see the problem. When you do only:
> 
> ```
> label_tc = stc.extract_label_time_course(fname_aseg, src=src, trans=inv['mri_head_t'])
> ```
> 
> It will only use the builtin FreeSurfer LUT with the `aseg` you have provided. You need to pass the `atlas_ids` you want to use with this non-standard MRI, e.g.:
> 
> ```
> atlas_ids, lut_colors = mne.read_freesurfer_lut('Schaefer2018_100Parcels_7Networks_LUT.txt')
> label_names = mne.get_volume_labels_from_aseg(fname_aseg, atlas_ids=atlas_ids)
> # not sure if you need this, but just to be safe:
> atlas_ids = {k: atlas_ids[k] for k in label_names}
> label_tc = stc.extract_label_time_course((fname_aseg, atlas_ids), src=src, trans=inv['mri_head_t'])
> ```

Ok well I see, thanks ! I tested it on another dataset and it's working fine. 👍

**Comment by larsoner** [comment]
Since @agramfort approves and it seems to be working for people, let's get this in so that people can start using it. Thanks for testing and iterating @poojaprabhu9 @vferat !

**Comment by poojaprabhu9** [comment]
stc.extract_label_time_course( labels, src, mode='auto', allow_empty=False,trans=None, mri_resolution=True, verbose=None)
We will be using the HCPMMP1 parcellation. This parcellation has atlas in .nii format. 
While passing the atlas path to the 'labels', is it mandatory to have the atlas in .mgz format?

**Comment by larsoner** [comment]
Yes right now, because everything is done in Freesurfer surface RAS coordinates (`vox_mri_t` in MNE, `get_vox2ras_tkr` in nibabel) and this doesn't exist AFAIK for NIfTI. Right now you have to make sure that whatever atlas image you use has been transformed properly to the subject's MRI space so that things like `mri_resolution=True` work properly, and that a given voxel in your atlas corresponds to the same voxel in your `T1.mgz`. The builtin Freesurfer `aseg.mgz` meets these requirements.

Have you already transformed the HCPMMP1 parcellation to your subject's MRI space such that a given voxel in your HCPMMP1 volume corresponds to the same voxel in their `aseg.mgz` / `T1.mgz`? If so, how did you do this transformation (we should document it)? If on the other hand you just have HCPMMP1 in NIfTI for fsaverage, we'll need to figure out how to transform it to the subject's space in the first place. It should be doable, but we need to figure it out. Once we figure that out, we can document it and then figure out what needs to be done at the MNE end to facilitate using other atlases.

**Comment by poojaprabhu9** [comment]
Check this link: https://cjneurolab.org/2016/11/22/hcp-mmp1-0-volumetric-nifti-masks-in-native-structural-space/
As per the link, the code to create volume parcellation using HCPMMP1 is [here](https://gist.github.com/poojaprabhu9/81bd61e081fc52c17392ffaaf0673e7a). the final volume will be in .nii format.
We just need it for fsaverage brain. 
We need to apply this parcellation on the already morphed volume estimates. For this purpose we have to extract the time courses for each labels of HCPMMP1 parcellation.  Currently, to extract the time courses we have the atlas in nifti. So the previous question.

**Comment by larsoner** [comment]
So `fsaverage` is pretty nice:
```
>>> import nibabel as nib
>>> img = nib.load('/mnt/bakraid/data/structurals/fsaverage/mri/aseg.mgz')
>>> import numpy as np
>>> np.array_equal(img.affine, img.header.get_vox2ras_tkr())
True
```
You don't have to worry about these differences. So in theory if you just save your `nii` to `mgz`, the MNE code should work, because (I think) `nibabel` will write the `vox2ras_tkr` as being equivalent to `vox2ras` / `affine`.

**Comment by larsoner** [comment]
... and personally I would probably update that script you used to output .mgz files with the vox2ras_tkr set properly. Then it would even work for non-fsaverage subjects.

**Comment by poojaprabhu9** [comment]
You mean to say I can use atlas in .nii format to extract time courses from stc?
(Without updating the MNE)

**Comment by larsoner** [comment]
> Or you mean I need to convert .nii to .mgz? Then try to extract the time course

You almost certainly need to convert it to `mgz`. Even just doing `mri_convert /path/whatever.nii /path/whatever.mgz` might be enough for things to work.

**Comment by virvw** [comment]
Hi there,

I am understanding Pooja wishes to extract volumic parcels using the
HCPMMP1 for our source analyses.
We are slowly advancing. Bear in mind you are not talking to coders but
simple scientists trying to make whatever receipe you guys mention work :)

I ran 45 min of a code to realize that "wm.mgz" was missing in the
fsaverage. It should be mentionned beforehand.
So before I continue solving error after bug sequentially: where can I get
a sense of the ingredients that are first needed to get the final receipe
to work?

------------------------------------------------
Right now, I came up with a rather unsatiscatory recipe because I think a
lot of initial variables are missing:

[This procedure takes 45 min +]

To create HCPMMP volumetric parcels on NeuroSpin server:
1) copy "fsaverage" in the MRI folder of your study
2) copy "lh.HCPMMP1.annot" and "rh.HCPMMP1.annot" in /fsaverage/label/
https://cjneurolab.org/2016/11/22/hcp-mmp1-0-volumetric-nifti-masks-in-native-structural-space/

3) create a "subject_list.txt" simply listing "fsaverage"
4) type to initialize freesurfer:
freesurfer_init
5) type to temporarily set your default $SUBJECTS_DIR to your study:
export $SUBJECTS_DIR=/neurospin/meg/meg_tmp/mystudy/MRI
6) type to inialize fsl  :
fsl_init
7) type to create the parcel into the folder HCPMMP1_parcellation:
bash create_subj_volume_parcellation.sh -L subject_list.txt -a HCPMMP1 -d
HCPMMP1_parcellation -s YES -t YES
---------------------------------------

Thanks a lot, much appreciated


Virginie van Wassenhove
*https://brainthemind.com/ <https://brainthemind.com/>*

**Comment by larsoner** [comment]
We can probably ship some HCPMMP1 + aseg volumes for `fsaverage` at least. I think that doing the conversion is out of scope for us. But it sounds like having the volumes for `fsaverage` is already enough, assuming you're morphing all subject data there anyway (?).

**Comment by larsoner** [comment]
... looking into this a bit, the solution they posted is a start but it's not exactly straightforward to get it working. I trimmed the script a bit just to use with fsaverage and output an mgz file. However, it does not output the actual region LUT that is to be used with the volume, so this needs to be created based on the math it actually does. And doing these things in Bash is not exactly straightforward.

I think it would actually be easier to do most of this in Python with nibabel. All that we really need FreeSurfer for (I think) is to generate a aseg volume from the annotation:
```
mri_aparc2aseg --s fsaverage --o HCPMMP1.mgz  --annot HCPMMP1
```
From there in Python we can:

1. Load `HCPMMP1.mgz`
2. Remove hippocampal voxels (they do this in the script)
3. Shift the values by some amount (they do this in the script)
4. Load the standard `aseg.mgz` for the subject, and add it to the HCPMMP one (including only whatever regions should be included)
5.  Keep track of which vertex values correspond to which regions, and then write out a custom LUT.

Eventually this could be done for other subjects, too, for example by using `read_labels_from_annot` followed by `morph_labels` then `write_labels_to_annot` for each subject (they do this sort of operation in their bash script).

For now, perhaps try using `aparc.a2009s+aseg.mgz`. Each FreeSurfer-reconstructed subject should already have this. It's a much less challenging path, and you should already have the files you need. In principle later HCPMMP1 can be substituted as the atlas+LUT combination, but in the meantime you can set up your scripts to work without this being a bottleneck.

**Comment by larsoner** [comment]
@poojaprabhu9 @virvw can you try this FreeSurfer command (after running `mne.datasets.fetch_hcpmmp_parcellation`):
```
$ mri_aparc2aseg --s fsaverage --volmask --annot HCPMMP1
$ mri_aparc2aseg --s fsaverage --volmask --annot HCPMMP1_combined
```
Then run [this gist](https://gist.github.com/larsoner/8e664205cd8285ca7c46211403ad12ce), which will create `SUBJECTS_DIR/fsaverage/mri/HCPMMP1ColorLUT.txt` and `HCPMMP1_combinedColorLUT.txt`. You can then view these with FreeSurfer (from `SUBJECTS_DIR/fsaverage/mri` directory:
```
$ freeview -v HCPMMP1+aseg.mgz:colormap=lut:lut=HCPMMP1ColorLUT.txt
```
Screenshot from 2020-05-08 12-30-43
```
freeview -v HCPMMP1_combined+aseg.mgz:colormap=lut:lut=HCPMMP1_combinedColorLUT.txt
```
Screenshot from 2020-05-08 12-42-29

Then in MNE you can do something like the following to extract label time courses:
```
mri_dir = op.join(mne.get_config('SUBJECTS_DIR'), 'fsaverage', 'mri')
labels = mne.read_labels_from_annot('fsaverage', 'HCPMMP1', 'both')
lut = mne.read_freesurfer_lut(op.join(mri_dir, 'HCPMMP1ColorLUT.txt'))
# limit to just the HCPMMP labels with this, or omit it to keep subcortical ones:
lut = {label.name: lut[label.name] for label in labels}
atlas = op.join(mri_dir, 'HCPMMP1+aseg.mgz')
label_tc = mne.extract_label_time_course(stc, (atlas, lut), fwd['src'])
```

**Comment by poojaprabhu9** [comment]
I was trying to use the above code but unfortunately i could not read the stc file (i updated the mne using, pip install https://api.github.com/repos/mne-tools/mne-python/zipball/master) because of this error.
TypeError: vertices must be an instance of list, got <class 'numpy.ndarray'> instead
I remember this problem was solved (when i was using, pip install -U https://github.com/larsoner/mne-python/archive/atlas.zip)
Did i miss something?

**Comment by larsoner** [comment]
> pip install https://api.github.com/repos/mne-tools/mne-python/zipball/master

This might not have done anything, try doing it again with `--upgrade` in there. If it still fails, upload the STC somewhere and I can see if I can load it

**Comment by poojaprabhu9** [comment]
I tried upgrading by 
`pip --upgrade install https://api.github.com/repos/mne-tools/mne-python/zipball/master`
Still the error exist.

You can find the stc file [here](https://mab.to/wp1AuZW4D)
Please try` stc.in_label` and `stc.extract_time_courses`.

**Comment by larsoner** [comment]
Can you upload the source space that you used? I can't try those functions without it

**Comment by poojaprabhu9** [comment]
just run this line
```
src = mne.setup_volume_source_space(
                        'fsaverage', pos=5., subjects_dir=subjects_dir, sphere=(0,0,0,90),
                        verbose=True,sphere_units='mm')
```

**Comment by larsoner** [comment]
Okay -- it's better to use the bem/brain to define the set of sources to keep rather than `sphere`

**Comment by poojaprabhu9** [comment]
> Okay -- it's better to use the bem/brain to define the set of sources to keep rather than `sphere`

will try that for the further computations.

**Comment by virvw** [comment]
@larsoner @poojaprabhu9 
This worked:
$ mri_aparc2aseg --s fsaverage --volmask --annot HCPMMP1
$ mri_aparc2aseg --s fsaverage --volmask --annot HCPMMP1_combined

I am stuck with the gist for several reasons independent of the code itself
@poojaprabhu9 if you manage to run that bit on the fsaverage and forward the outcome, I'll finish up...

**Comment by virvw** [comment]
@larsoner @poojaprabhu9 

The gist was ran removing the initial option "sort" because not available in our version.
 
Note that the created file "FreeSurferColorLUT.txt" was found by the script "create_subj_volume_parcellation.sh" when placed in the same folder as the script and not in the "fsaverage/label/"  

Now I am encountering issue when running "create_subj_volume_parcellation.sh" since we are missing "wm.mgz" for fsaverage


-----------------
FreeSurfer/vv221713@is152869:/neurospin/meg/meg_tmp/2015_MiniManip_rsMEG/POLTI_Ignacio_RSretroTIME/MRI$ bash create_subj_volume_parcellation.sh -L subject_list.txt -a HCPMMP1 -d HCPMMP1_parcellation -s YES -t YES

         >>>>         Current FreeSurfer subjects folder is /neurospin/meg/meg_tmp/2015_MiniManip_rsMEG/POLTI_Ignacio_RSretroTIME/MRI

reading colortable from annotation file...
colortable with 181 entries read (originally ./left.fsaverage164.label.gii)
reading colortable from annotation file...
colortable with 181 entries read (originally ./left.fsaverage164.label.gii)
reading colortable from annotation file...
colortable with 181 entries read (originally ./right.fsaverage164.label.gii)
reading colortable from annotation file...
colortable with 181 entries read (originally ./right.fsaverage164.label.gii)

         >>>>         PREPROCESSING fsaverage         <<<<
         >>>>         START TIME: Sun May 10 16:49:56 CEST 2020         <<<<
>>>>    Annotation files lh.fsaverage_HCPMMP1.annot and rh.fsaverage_HCPMMP1.annot already exist in fsaverage/label. Won't perform transformations
reading colortable from annotation file...
colortable with 182 entries read (originally HCPMMP1_parcellation/temp_1_0_1246/colortab_HCPMMP1_L)
reading colortable from annotation file...
colortable with 182 entries read (originally HCPMMP1_parcellation/temp_1_0_1246/colortab_HCPMMP1_R)
fslmaths: error while loading shared libraries: libopenblas.so.0: cannot open shared object file: No such file or directory
fslmaths: error while loading shared libraries: libopenblas.so.0: cannot open shared object file: No such file or directory
fslmaths: error while loading shared libraries: libopenblas.so.0: cannot open shared object file: No such file or directory
fslmaths: error while loading shared libraries: libopenblas.so.0: cannot open shared object file: No such file or directory
fslmaths: error while loading shared libraries: libopenblas.so.0: cannot open shared object file: No such file or directory
>> Creating subcortical aseg masks for subject fsaverage
fslmaths: error while loading shared libraries: libopenblas.so.0: cannot open shared object file: No such file or directory
fslmaths: error while loading shared libraries: libopenblas.so.0: cannot open shared object file: No such file or directory
fslmaths: error while loading shared libraries: libopenblas.so.0: cannot open shared object file: No such file or directory
fslmaths: error while loading shared libraries: libopenblas.so.0: cannot open shared object file: No such file or directory
fslmaths: error while loading shared libraries: libopenblas.so.0: cannot open shared object file: No such file or directory
fslmaths: error while loading shared libraries: libopenblas.so.0: cannot open shared object file: No such file or directory
fslmaths: error while loading shared libraries: libopenblas.so.0: cannot open shared object file: No such file or directory
fslmaths: error while loading shared libraries: libopenblas.so.0: cannot open shared object file: No such file or directory
fslmaths: error while loading shared libraries: libopenblas.so.0: cannot open shared object file: No such file or directory
fslmaths: error while loading shared libraries: libopenblas.so.0: cannot open shared object file: No such file or directory
fslmaths: error while loading shared libraries: libopenblas.so.0: cannot open shared object file: No such file or directory
fslmaths: error while loading shared libraries: libopenblas.so.0: cannot open shared object file: No such file or directory
computing statistics for each annotation in /neurospin/meg/meg_tmp/2015_MiniManip_rsMEG/POLTI_Ignacio_RSretroTIME/MRI/fsaverage/label/lh.fsaverage_HCPMMP1.annot.
reading volume /neurospin/meg/meg_tmp/2015_MiniManip_rsMEG/POLTI_Ignacio_RSretroTIME/MRI/fsaverage/mri/wm.mgz...
mghRead(/neurospin/meg/meg_tmp/2015_MiniManip_rsMEG/POLTI_Ignacio_RSretroTIME/MRI/fsaverage/mri/wm.mgz, -1): could not open file
mris_anatomical_stats: could not read input volume /neurospin/meg/meg_tmp/2015_MiniManip_rsMEG/POLTI_Ignacio_RSretroTIME/MRI/fsaverage/mri/wm.mgz
computing statistics for each annotation in /neurospin/meg/meg_tmp/2015_MiniManip_rsMEG/POLTI_Ignacio_RSretroTIME/MRI/fsaverage/label/rh.fsaverage_HCPMMP1.annot.
reading volume /neurospin/meg/meg_tmp/2015_MiniManip_rsMEG/POLTI_Ignacio_RSretroTIME/MRI/fsaverage/mri/wm.mgz...
mghRead(/neurospin/meg/meg_tmp/2015_MiniManip_rsMEG/POLTI_Ignacio_RSretroTIME/MRI/fsaverage/mri/wm.mgz, -1): could not open file
mris_anatomical_stats: could not read input volume /neurospin/meg/meg_tmp/2015_MiniManip_rsMEG/POLTI_Ignacio_RSretroTIME/MRI/fsaverage/mri/wm.mgz
(standard_in) 2: syntax error
sed: -e expression n°1, caractère 3: `,' inattendue
sed: -e expression n°1, caractère 3: `,' inattendue
sed: -e expression n°1, caractère 3: `,' inattendue
sed: -e expression n°1, caractère 3: `,' inattendue

         >>>>         fsaverage STARTED AT Sun May 10 16:49:56 CEST 2020, ENDED AT: Sun May 10 16:51:04 CEST 2020

**Comment by larsoner** [comment]
> The gist was ran removing the initial option "sort" because not available in our version.

It will not work properly without that option. The regions will be mislabeled. It was added to `master` recently, so if you update it should be available.

> Note that the created file "FreeSurferColorLUT.txt" was found by the script

You should not run this at all anymore, just [the two steps mentioned in the comment above](https://github.com/mne-tools/mne-python/pull/7639#issuecomment-625907891):

1. `$ mri_aparc2aseg --s fsaverage --volmask --annot HCPMMP1`
2. `python hcpmmp_lut.py` from [here](https://gist.github.com/larsoner/8e664205cd8285ca7c46211403ad12ce)

This should create all the files for `fsaverage` that you need to be able to extract label time courses (i.e., run the Python code from [this comment](https://github.com/mne-tools/mne-python/pull/7756#issuecomment-625986409)).

**Comment by poojaprabhu9** [comment]
> python hcpmmp_lut.py from here

There is something missing in the line 17 of hcpmmp_lut.py. is it so?

**Comment by virvw** [comment]
@larsoner considering thet the sort option is set to FALSE in your script,
does that matter at all?

If it does then I have to wait unless someone can run your script because I
can't update.

@pooja due to this issue, please neglect the files I just sent you which
will be mislabeled...

Virginie van Wassenhove
*https://brainthemind.com/ <https://brainthemind.com/>*

**Comment by larsoner** [comment]
> There is something missing in the line 17 of hcpmmp_lut.py. is it so?

That's the start of a multi-line string, looks okay to me. Did you try and and get some error?

> considering thet the sort option is set to FALSE in your script,
does that matter at all?

In `master` the default is `True` (and anything before `master` implicitly had `sort=True`), so yes unfortunately you need it.

Also, if you plan to use the `HCPMMP1_combined` version, you probably need to (for now at least) regenerate it using latest `master`, which you can do by deleting the `lh/rh.HCPMMP1_combined.annot` files from `fsaverage/label` the doing `mne.datasets.fetch_hcpmmp_parcellation()`. And if you already did the FreeSurfer `mri_aparc2aseg` command, redo it as well, and then re-run my `gist`. This also has to do with the `sort` option defauting to True, which causes problems when generating the low-res/combined parc...

**Comment by poojaprabhu9** [comment]
> That's the start of a multi-line string, looks okay to me. Did you try and and get some error?
Everything is fine. It was a mistake from my side.

**Comment by virvw** [comment]
Ok I think things are sorted out.... and it is now back in Pooja's hands...!


Virginie van Wassenhove
*https://brainthemind.com/ <https://brainthemind.com/>*

**Comment by poojaprabhu9** [comment]
@larsoner @virvw 
It is working!!
Thanks a lot @larsoner

**Pull request merged**